### PR TITLE
Generate: add config-level validation

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

This PR adds generation argument validation that can be performed at a `generation_config` level. 

It aims at reducing the number of issues due to incorrect parameterization, where the user expects some argument to modify the output but it doesn't.

### CODE REVIEW

File: src/transformers/generation/configuration_utils.py [hunk 1]
```diff
@@ -332,12 +332,121 @@ def __repr__(self):
...
+        Note that some parameters are best validated at generate runtime, as they may depend on other inputs and/or the
+        model, such as parameters related to the generation length.
         """
+
+        # Validation of individual attributes
         if self.early_stopping not in {True, False, "never"}:
             raise ValueError(f"`early_stopping` must be a boolean or 'never', but is {self.early_stopping}.")
 
+        # Validation of attribute relations:
+        # 1. detect sampling-only parameterization when not in sampling mode
```
Comment (gante) [inline src/transformers/generation/configuration_utils.py:347]: The validation in `1.` and `2.` is new, `3.` and `4.` were moved from `generate` (as it falls under `generation_config` level validation)

The new validation checks throw exceptions, as they correspond to an impossible combination of arguments (and most users ignore warnings). Happy to lower them to warnings, if you prefer to optimize for backwards compatibility :)

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -1651,12 +1633,6 @@ def generate(
             )
 
         elif generation_mode == GenerationMode.BEAM_SEARCH:
-            if generation_config.num_return_sequences > generation_config.num_beams:
-                raise ValueError("`num_return_sequences` has to be smaller or equal to `num_beams`.")
-
-            if stopping_criteria.max_length is None:
```
Comment (gante) [inline src/transformers/generation/utils.py:1657]: This exception can't be reached: we always set `max_length` before initializing `stopping_criteria`, which means we will always have a `max_length`-related stopping criteria.

File: src/transformers/generation/utils.py [hunk 3]
```diff
@@ -1651,12 +1633,6 @@ def generate(
...
-                raise ValueError("`max_length` needs to be a stopping_criteria for now.")
```
Comment (gante) [inline src/transformers/generation/utils.py:1658]: This exception can't be reached: we always set max_length before initializing stopping_criteria, which means we will always have a max_length-related stopping criteria.

**Review by sgugger — APPROVED** [review]
Thanks for cleaning this up!

**Comment by fxmarty** [comment]
@gante 

The warnings
```
/home/felix/transformers/src/transformers/generation/configuration_utils.py:412: UserWarning: `do_sample` is set to `False`. However, `top_p` is set to `0.6` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `top_p`.
  warnings.warn(
/home/felix/transformers/src/transformers/generation/configuration_utils.py:422: UserWarning: `do_sample` is set to `False`. However, `top_k` is set to `1` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `top_k`.
  warnings.warn(
```

do not explicit how to unset the parameters. My first intuition was `top_k=None`, etc., but this does not work. In fact, one needs to call specifically:

```
gen_out = model.generate(**inputs, do_sample=False, temperature=1, top_p=1, top_k=50)
```

to remove the warnings, which I find counter-intuitive. Could we allow `None`?

**Comment by gante** [comment]
@fxmarty https://github.com/huggingface/transformers/pull/29119 :D

**Pull request merged**